README.md
# push-receiver

A library to subscribe to GCM/FCM and receive notifications within a node process.

For [Electron](https://github.com/electron/electron), you can use [electron-push-receiver](https://github.com/MatthieuLemoine/electron-push-receiver) instead which provides a convenient wrapper.

See [this blog post](https://medium.com/@MatthieuLemoine/my-journey-to-bring-web-push-support-to-node-and-electron-ce70eea1c0b0) for more details.

## When should I use `push-receiver` ?

- I want to **receive** push notifications sent using Firebase Cloud Messaging in an [electron](https://github.com/electron/electron) desktop application.
- I want to communicate with a node process/server using Firebase Cloud Messaging infrastructure.

## When should I not use `push-receiver` ?

- I want to **send** push notifications (use the firebase SDK instead)
- My application is running on a FCM supported platform (Android, iOS, Web).

## Install

`
npm i -S @tmbill/push-receiver
`

## Requirements

- Node v16 (async/await/randomUUID support)
- Firebase sender id to receive notification
- Firebase serverKey to send notification (optional)


## Usage

### ClientConfig

```typescript
interface ClientConfig {
    credentials?: Credentials // Will be generated if missing - save this after first use!
    persistentIds?: PersistentId[] // Default - []
    senderId: string // Required
    bundleId?: string // Default - 'receiver.push.com'
    chromeId?: string // Default - 'org.chromium.linux'
    chromeVersion?: string // Default - '94.0.4606.51'
    skipFcmRegistration?: boolean // Default - false
    logLevel?: keyof typeof LogLevels // 'NONE'|'DEBUG'|'VERBOSE' - default: 'NONE'
    vapidKey?: string // Default - default firebase VAPID key
    heartbeatIntervalMs?: number // Default - 5 * 60 * 1000
}
```

### Node example

```javascript
import { PushReceiver } from '@tmbill/push-receiver'
import { argv as parsedArgs } from 'yargs'

if (!parsedArgs.senderId) {
    console.error('Missing senderId')
    return
}

(async () => {
    const instance = new PushReceiver({
        logLevel: parsedArgs.logLevel || 'DEBUG',
        senderId: parsedArgs.senderId,
        persistentIds: [], // Recover stored ids of all previous notifications
    })

    const stopListeningToCredentials = instance.onCredentialsChanged(({ oldCredentials, newCredentials }) => {
        console.log('Client generated new credentials.', newCredentials)
        // Save them somewhere! And decide if thing are needed to re-subscribe
    })

    const stopListeningToNotifications = instance.onNotification(notification => {
        // Do someting with the notification
        console.log('Notification received', notification)
    })

    await instance.connect()

    if (parsedArgs.serverId) {
        await instance.testMessage(parsedArgs.serverId)
    }

    stopListeningToCredentials()
    stopListeningToNotifications()
    instance.destroy()
})()
```


CHANGELOG.md
# @tmbill/push-receiver

## 3.1.0

### Minor Changes

- Added automated Heartbeat messages
  - new option `heartbeatIntervalMs` DEFAULT: 5 * 60 * 1000
  - new events `ON_HEARTBEAT` - this is emited when socket recieves `ping` or `ack` messages

### Patch Changes

- Updated devDependencies

test/4kb.js
/* eslint-disable no-undef */
module.exports =
  "iVBORw0KGgoAAAANSUhEUgAAAAwAAAACCAYAAABsfz2XAAAMFGlDQ1BJQ0MgUHJvZmlsZQAASImVVwdUk8kWnr+kEBJaINIJvQnSq0DoIAhIBxshCRBKgISgYkcXFVy7WFBUdAVE0bUAshZE7CyCvT8QUVlZFws2VN6kgK6vnXfPmX++3Ln3zncnd+bMAKBsx87Ly0ZVAMgRFAijg/2YiUnJTFIPwIAe0AA6QIvNEeX5RkWFAyij/d/l3S2ASPrrNpJY/zr+X0WVyxNxAECiIE7lijg5EB8FANfk5AkLACC0Q73xrII8CR6EWF0ICQJAxCU4XYY1JThVhsdLbWKj/SFmAUCmstnCdACUJLyZhZx0GEdJwtFOwOULIN4GsTcng82F+AHE43NyciFWJkNskfpdnPS/xUwdi8lmp49hWS5SIQfwRXnZ7Dn/53L8b8nJFo/OYQQbNUMYEi3JGa5bTVZumARTIT4hSI2IhFgN4ot8rtRegu9liEPi5PYDHJE/XDPAAAAFXHZAGMS6EDPEWXG+cuzAFkp9oT0awS8IjZXjVGFutDw+WijIjgiXx1mewQsdxZU8UWDMqE0aPygUYlhp6NGijNgEGU+0rZAfHwGxEsSdoqyYMLnvo6IM/4hRG6E4WsLZBOK3acKgaJkNppkjGs0Ls+WwpXPBWsBYBRmxITJfLJEnSgwf5cDlBQTKOGBcniBOzg2D1eUXLfctycuOkttjlbzs4GjZOmOHRIUxo77XCmCBydYBe5zJnhQln+tdXkFUrIwbjoJw4A8CABOIYUsFuSAT8DsGGgfgL9lIEGADIUgHPGAj14x6JEhHBPAbA4rAnxDxgGjMz086ygOFUP9lTCv72oA06Wih1CMLPIU4B9fGvXFPPBx+WbA54G64+6gfU3l0VmIgMYAYQgwiWo7x4EDW2bAJAf/f6MJgz4PZSbgIRnP4Fo/wlNBFeEy4Segm3AXx4Ik0itxqJr9Y+ANzJpgMumG0IHl2qd9nh5tB1s64H+4F+UPuOAPXBja4E8zEF/eBuTlD7fcMxWPcvq3lj/NJWH+fj1yvZKXkLGeROvbP+I9Z/RjF/7s14sI+7EdLbDl2BLuAncEuYSewRsDETmNNWDt2UoLHKuGJtBJGZ4uWcsuCcfijNnZ1dv12n3+Ymy2fX7JeogLe7ALJZvDPzZsj5KdnFDB94WnMY4YKOLbjmQ529q4ASM522dHxhiE9sxHG5W+6/BYA3EuhMv2bjm0MwPGnANDffdMZv4blvgaAk50csbBQppMcx4AAKEAZ7gotoA+MgQXMxwG4AE/AAoFgEogEsSAJzIArngFyIOdZYB5YDEpAGVgDNoKtYAfYDWrAAXAYNIIT4Aw4D66ATnAT3Id10QdegEHwDgwjCEJCaAgd0UIMEFPEGnFA3BBvJBAJR6KRJCQFSUcEiBiZhyxBypB1yFZkF1KL/IocR84gl5Au5C7Sg/Qjr5FPKIZSUXVUDzVDJ6BuqC8ahsai09F0NB8tQpeiq9DNaBW6H21Az6BX0JtoN/oCHcIApogxMEPMBnPD/LFILBlLw4TYAqwUK8eqsHqsGf7P17FubAD7iBNxOs7EbWBthuBxOAfPxxfgK/GteA3egLfh1/EefBD/SqARdAnWBA9CKCGRkE6YRSghlBP2Eo4RzsF900d4RyQSGURzoivcl0nETOJc4kriduJBYguxi9hLHCKRSFoka5IXKZLEJhWQSkhbSPtJp0nXSH2kD2RFsgHZgRxETiYLyMXkcvI+8inyNfIz8rCCioKpgodCpAJXYY7CaoU9Cs0KVxX6FIYpqhRzihcllpJJWUzZTKmnnKM8oLxRVFQ0UnRXnKLIV1ykuFnxkOJFxR7Fj1Q1qhXVnzqNKqauolZTW6h3qW9oNJoZjUVLphXQVtFqaWdpj2gflOhKtkqhSlylhUoVSg1K15ReKisomyr7Ks9QLlIuVz6ifFV5QEVBxUzFX4WtskClQuW4ym2VIVW6qr1qpGqO6krVfaqXVJ+rkdTM1ALVuGpL1XarnVXrpWN0Y7o/nUNfQt9DP0fvUyeqm6uHqmeql6kfUO9QH9RQ03DSiNeYrVGhcVKjm4ExzBihjGzGasZhxi3Gp3F643zH8catGFc/7tq495o6mixNnmap5kHNm5qftJhagVpZWmu1GrUeauPaVtpTtGdpV2qf0x7QUdfx1OHolOoc1rmni+pa6UbrztXdrduuO6Snrxesl6e3Re+s3oA+Q5+ln6m/Qf+Ufr8B3cDbgG+wweC0wR9MDaYvM5u5mdnGHDTUNQwxFBvuMuwwHDYyN4ozKjY6aPTQmGLsZpxmvMG41XjQxMBkssk8kzqTe6YKpm6mGaabTC+YvjczN0swW2bWaPbcXNM81LzIvM78gQXNwsci36LK4oYl0dLNMstyu2WnFWrlbJVhVWF11Rq1drHmW2+37hpPGO8+XjC+avxtG6qNr02hTZ1Njy3DNty22LbR9uUEkwnJE9ZOuDDhq52zXbbdHrv79mr2k+yL7ZvtXztYOXAcKhxuONIcgxwXOjY5vnKyduI5VTrdcaY7T3Ze5tzq/MXF1UXoUu/S72rimuK6zfW2m7pblNtKt4vuBHc/94XuJ9w/erh4FHgc9vjL08Yzy3Of5/OJ5hN5E/dM7PUy8mJ77fLq9mZ6p3jv9O72MfRh+1T5PGYZs7isvaxnvpa+mb77fV/62fkJ/Y75vff38J/v3xKABQQHlAZ0BKoFxgVuDXwUZBSUHlQXNBjsHDw3uCWEEBIWsjbkdqheKCe0NnRwkuuk+ZPawqhhMWFbwx6HW4ULw5sno5MnTV4/+UGEaYQgojESRIZGro98GGUelR/12xTilKgpFVOeRttHz4u+EEOPmRmzL+ZdrF/s6tj7cRZx4rjWeOX4afG18e8TAhLWJXQnTkicn3glSTuJn9SUTEqOT96bPDQ1cOrGqX3TnKeVTLs13Xz67OmXZmjPyJ5xcqbyTPbMIymElISUfSmf2ZHsKvZQamjqttRBjj9nE+cFl8XdwO3nefHW8Z6leaWtS3ue7pW+Pr0/wyejPGOA78/fyn+VGZK5I/N9VmRWddZIdkL2wRxyTkrOcYGaIEvQlqufOzu3K886rySvO98jf2P+oDBMuFeEiKaLmgrU4TWnXWwh/kncU+hdWFH4YVb8rCOzVWcLZrfPsZqzYs6zoqCiX+biczlzW+cZzls8r2e+7/xdC5AFqQtaFxovXLqwb1HwoprFlMVZi38vtiteV/x2ScKS5qV6Sxct7f0p+Ke6EqUSYcntZZ7LdizHl/OXd6xwXLFlxddSbunlMruy8rLPKzkrL/9s//Pmn0dWpa3qWO2yunINcY1gza21Pmtr1qmuK1rXu37y+oYNzA2lG95unLnxUrlT+Y5NlE3iTd2bwzc3bTHZsmbL560ZW29W+FUc3Ka7bcW299u5269Vsirrd+jtKNvxaSd/551dwbsaqsyqyncTdxfufronfs+FX9x+qd2rvbds75dqQXV3TXRNW61rbe0+3X2r69A6cV3//mn7Ow8EHGiqt6nfdZBxsOwQOCQ+9MevKb/eOhx2uPWI25H6o6ZHtx2jHyttQBrmNAw2ZjR2NyU1dR2fdLy12bP52G+2v1WfMDxRcVLj5OpTlFNLT42cLjo91JLXMnAm/Uxv68zW+2cTz95om9LWcS7s3MXzQefPXvC9cPqi18UTlzwuHb/sdrnxisuVhnbn9mO/O/9+rMOlo+Gq69WmTvfO5q6JXaeu+Vw7cz3g+vkboTeu3Iy42XUr7tad29Nud9/h3nl+N/vuq3uF94bvL3pAeFD6UOVh+SPdR1X/sPzHwW6X7pM9AT3tj2Me3+/l9L54InryuW/pU9rT8mcGz2qfOzw/0R/U3/nH1D/6XuS9GB4o+";


test/keys.template.js
/* eslint-disable no-undef */
// Set a sender id and server key and save this under test/keys.js
module.exports = {
  SENDER_ID: "Your sender id here",
  SERVER_KEY: "Your server key here",
};


test/notification.test.js
/* eslint-disable no-undef */
/* eslint-disable @typescript-eslint/no-var-requires */
const { SENDER_ID, SERVER_KEY } = require("./keys");
const Client = require("../dist/client.js").default;

const NOTIFICATIONS = {
  SIMPLE: { title: "Hello world ", body: "Test" },
  LARGE: { title: "Hello world ", body: require("./4kb") },
};

const client = new Client({
  senderId: SENDER_ID,
  logLevel: "DEBUG",
  persistentIds: [],
  heartbeatIntervalMs: 10000,
});

let credentials;

client.onCredentialsChanged((credentialsChange) => {
  credentials = credentialsChange.newCredentials;
});

jest.setTimeout(20000);

describe("Parser", function () {
  beforeAll(async function () {
    await client.connect();
  });

  it("should receive a simple notification", async function () {
    send(NOTIFICATIONS.SIMPLE);
    const notifications = await receive(1);
    expect(notifications.length).toEqual(1);
    expect(notifications[0].message.data).toEqual(NOTIFICATIONS.SIMPLE);
  });

  it("should receive a large notification", async function () {
    send(NOTIFICATIONS.LARGE);
    const notifications = await receive(1);
    expect(notifications.length).toEqual(1);
    expect(notifications[0].message.data).toEqual(NOTIFICATIONS.LARGE);
  });

  afterAll(() => {
    client.setLogLevel("NONE");
    client.destroy();
  });
});

async function send(notification) {
  try {
    await client.send(notification, SERVER_KEY);
  } catch (e) {
    throw new Error(`sending of notification failed: ${e}`);
  }
}

async function receive(n) {
  const received = [];

  return new Promise((resolve) => {
    client.onNotification((notification) => {
      console.log("got notification", notification);

      received.push(notification);

      if (received.length >= n) {
        resolve(received);
      }
    });
  });
}


src/fcm.ts
import crypto from "crypto";
import request from "./utils/request";
import { escape } from "./utils/base64";

import type * as Types from "./types";

const FCM_SUBSCRIBE = "https://fcm.googleapis.com/fcm/connect/subscribe";
const FCM_ENDPOINT = "https://fcm.googleapis.com/fcm/send";
// const FCM_ENDPOINT =
//   "https://fcm.googleapis.com/v1/projects/tmbillchannel/messages:send";

export default async function registerFCM(
  gcm: Types.GcmData,
  config: Types.ClientConfig
): Promise<Types.Credentials> {
  const keys = await createKeys();
  const response = await request<Types.FcmData>({
    ...config.axiosConfig,
    url: FCM_SUBSCRIBE,
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    data: new URLSearchParams({
      authorized_entity: config.senderId,
      endpoint: `${FCM_ENDPOINT}/${gcm.token}`,
      encryption_key: keys.publicKey
        .replace(/=/g, "")
        .replace(/\+/g, "-")
        .replace(/\//g, "_"),
      encryption_auth: keys.authSecret
        .replace(/=/g, "")
        .replace(/\+/g, "-")
        .replace(/\//g, "_"),
    }).toString(),
  });

  return {
    gcm,
    keys,
    fcm: response,
  };
}

function createKeys(): Promise<Types.Keys> {
  return new Promise((resolve, reject) => {
    const dh = crypto.createECDH("prime256v1");

    dh.generateKeys();
    crypto.randomBytes(16, (err, buf) => {
      if (err) {
        return reject(err);
      }

      return resolve({
        privateKey: escape(dh.getPrivateKey("base64")),
        publicKey: escape(dh.getPublicKey("base64")),
        authSecret: escape(buf.toString("base64")),
      });
    });
  });
}


src/client.ts
import { EventEmitter } from "events";
import Long from "long";
import ProtobufJS from "protobufjs";
import tls from "tls";
import axios from "axios";
import registerGCM, { checkIn } from "./gcm";
import registerFCM from "./fcm";
import Parser from "./parser";
import decrypt from "./utils/decrypt";
import Logger from "./utils/logger";
import Protos, { mcs_proto } from "./protos";

import { Variables, MCSProtoTag } from "./constants";

import type * as Types from "./types";

export { Types };

ProtobufJS.util.Long = Long;
ProtobufJS.configure();

const FCM_SEND_API = "https://fcm.googleapis.com/fcm/send";
// const FCM_SEND_API =
//   "https://fcm.googleapis.com/v1/projects/tmbillchannel/messages:send";

const HOST = "mtalk.google.com";
const PORT = 5228;
const MAX_RETRY_TIMEOUT = 15;

export default class PushReceiver extends EventEmitter {
  private config: Types.ClientConfig;
  private socket: tls.TLSSocket;
  private retryCount = 0;
  private retryTimeout: NodeJS.Timeout;
  private parser: Parser;
  private heartbeatTimer?: NodeJS.Timeout;
  private heartbeatTimeout?: NodeJS.Timeout;
  private streamId = 0;
  private lastStreamIdReported = -1;

  public persistentIds: Types.PersistentId[];

  constructor(config: Types.ClientConfig) {
    super();

    Logger.setLogLevel(config.logLevel ?? "NONE");

    Logger.verbose("constructor", config);

    this.config = {
      bundleId: "receiver.push.com",
      chromeId: "org.chromium.linux",
      chromeVersion: "94.0.4606.51",
      vapidKey:
        "BDOU99-h67HcA6JeFXHbSNMu7e2yNNu3RzoMj8TM4W88jITfq7ZmPvIM1Iv-4_l2LxQcYwhqby2xGpWwzjfAnG4", // This is default Firebase VAPID key
      persistentIds: [],
      heartbeatIntervalMs: 5 * 60 * 1000, // 5 min
      ...config,
    };

    this.persistentIds = this.config.persistentIds;
  }

  public setLogLevel(logLevel: Types.ClientConfig["logLevel"]) {
    Logger.setLogLevel(logLevel);
  }

  public on(
    event: "ON_MESSAGE_RECEIVED",
    listener: (data: Types.MessageEnvelope) => void
  ): this;
  public on(
    event: "ON_CREDENTIALS_CHANGE",
    listener: (data: Types.EventChangeCredentials) => void
  ): this;
  public on(event: "ON_CONNECT", listener: (data: void) => void): this;
  public on(event: "ON_DISCONNECT", listener: (data: void) => void): this;
  public on(event: "ON_READY", listener: (data: void) => void): this;
  public on(event: "ON_HEARTBEAT", listener: (data: void) => void): this;
  public on(event: unknown, listener: unknown): this {
    return EventEmitter.prototype.on.apply(this, [event, listener]);
  }

  public off(
    event: "ON_MESSAGE_RECEIVED",
    listener: (data: Types.MessageEnvelope) => void
  ): this;
  public off(
    event: "ON_CREDENTIALS_CHANGE",
    listener: (data: Types.EventChangeCredentials) => void
  ): this;
  public off(event: "ON_CONNECT", listener: (data: void) => void): this;
  public off(event: "ON_DISCONNECT", listener: (data: void) => void): this;
  public off(event: "ON_READY", listener: (data: void) => void): this;
  public off(event: "ON_HEARTBEAT", listener: (data: void) => void): this;
  public off(event: unknown, listener: unknown): this {
    return EventEmitter.prototype.off.apply(this, [event, listener]);
  }

  public emit(
    event: "ON_MESSAGE_RECEIVED",
    data: Types.MessageEnvelope
  ): boolean;
  public emit(
    event: "ON_CREDENTIALS_CHANGE",
    data: Types.EventChangeCredentials
  ): boolean;
  public emit(event: "ON_CONNECT"): boolean;
  public emit(event: "ON_DISCONNECT"): boolean;
  public emit(event: "ON_READY"): boolean;
  public emit(event: "ON_HEARTBEAT"): boolean;
  public emit(event: unknown, ...args: unknown[]): boolean {
    Logger.debug("emit", event, ...args);
    return EventEmitter.prototype.emit.apply(this, [event, ...args]);
  }

  public onNotification(
    listener: (data: Types.MessageEnvelope) => void
  ): Types.DisposeFunction {
    this.on("ON_MESSAGE_RECEIVED", listener);
    return () => this.off("ON_MESSAGE_RECEIVED", listener);
  }

  public onCredentialsChanged(
    listener: (data: Types.EventChangeCredentials) => void
  ): Types.DisposeFunction {
    this.on("ON_CREDENTIALS_CHANGE", listener);
    return () => this.off("ON_CREDENTIALS_CHANGE", listener);
  }

  public onReady(listener: () => void): Types.DisposeFunction {
    this.on("ON_READY", listener);
    return () => this.off("ON_READY", listener);
  }

  public connect = async (): Promise<void> => {
    if (this.socket) return;

    Logger.verbose("connect");
    if (this.config.credentials) {
      Logger.verbose("checkin");
      await this.checkIn();
    } else {
      Logger.warn("Missing credentials... Runing register.");
      const oldCredentials = this.config.credentials;
      const newCredentials = await this.register();
      this.emit("ON_CREDENTIALS_CHANGE", { oldCredentials, newCredentials });
      this.config.credentials = newCredentials;
      Logger.verbose("got credentials", newCredentials);
    }

    this.lastStreamIdReported = -1;

    this.socket = new tls.TLSSocket(null);
    this.socket.setKeepAlive(true);
    this.socket.on("connect", this.handleSocketConnect);
    this.socket.on("close", this.handleSocketClose);
    this.socket.on("error", this.handleSocketError);
    this.socket.connect({ host: HOST, port: PORT });

    this.parser = new Parser(this.socket);
    this.parser.on("message", this.handleMessage);
    this.parser.on("error", this.handleParserError);

    this.sendLogin();

    return new Promise((res) => {
      const dispose = this.onReady(() => {
        dispose();
        res();
      });
    });
  };

  public destroy = () => {
    clearTimeout(this.retryTimeout);
    this.clearHeartbeat();

    if (this.socket) {
      this.socket.off("close", this.handleSocketClose);
      this.socket.off("error", this.handleSocketError);
      this.socket.destroy();
      this.socket = null;
    }

    if (this.parser) {
      this.parser.off("error", this.handleParserError);
      this.parser.destroy();
      this.parser = null;
    }
  };

  public async register(): Promise<Types.Credentials> {
    const subscription = await registerGCM(this.config);
    return registerFCM(subscription, this.config);
  }

  public checkIn(): Promise<Types.GcmData> {
    return checkIn(this.config);
  }

  private clearHeartbeat() {
    clearTimeout(this.heartbeatTimer);
    this.heartbeatTimer = undefined;

    clearTimeout(this.heartbeatTimeout);
    this.heartbeatTimeout = undefined;
  }

  private startHeartbeat() {
    this.clearHeartbeat();

    if (!this.config.heartbeatIntervalMs) return;

    this.heartbeatTimer = setTimeout(
      this.sendHeartbeatPing.bind(this),
      this.config.heartbeatIntervalMs
    );
    this.heartbeatTimeout = setTimeout(
      this.socketRetry.bind(this),
      this.config.heartbeatIntervalMs * 2
    );
  }

  private handleSocketConnect = (): void => {
    this.retryCount = 0;
    this.emit("ON_CONNECT");
    this.startHeartbeat();
  };

  private handleSocketClose = (): void => {
    this.emit("ON_DISCONNECT");
    this.clearHeartbeat();
    this.socketRetry();
  };

  private handleSocketError = (err: Error): void => {
    Logger.error(err);
    // ignore, the close handler takes care of retry
  };

  private socketRetry() {
    this.destroy();
    const timeout = Math.min(++this.retryCount, MAX_RETRY_TIMEOUT) * 1000;
    this.retryTimeout = setTimeout(this.connect, timeout);
  }

  private getStreamId(): number {
    this.lastStreamIdReported = this.streamId;
    return this.streamId;
  }

  private newStreamIdAvailable(): boolean {
    return this.lastStreamIdReported != this.streamId;
  }

  private sendHeartbeatPing() {
    const heartbeatPingRequest: Record<string, unknown> = {};

    if (this.newStreamIdAvailable()) {
      heartbeatPingRequest.last_stream_id_received = this.getStreamId();
    }

    Logger.verbose("Heartbeat send pong", heartbeatPingRequest);

    const HeartbeatPingRequestType = Protos.mcs_proto.HeartbeatPing;
    const errorMessage = HeartbeatPingRequestType.verify(heartbeatPingRequest);

    if (errorMessage) {
      throw new Error(errorMessage);
    }

    const buffer =
      HeartbeatPingRequestType.encodeDelimited(heartbeatPingRequest).finish();

    Logger.verbose("HEARTBEAT sending PING", heartbeatPingRequest);

    this.socket.write(
      Buffer.concat([Buffer.from([MCSProtoTag.kHeartbeatPingTag]), buffer])
    );
  }

  private sendHeartbeatPong(object) {
    const heartbeatAckRequest: Record<string, any> = {};

    if (this.newStreamIdAvailable()) {
      heartbeatAckRequest.last_stream_id_received = this.getStreamId();
    }

    if (object?.status) {
      heartbeatAckRequest.status = object.status;
    }

    Logger.verbose("Heartbeat send pong", heartbeatAckRequest);

    const HeartbeatAckRequestType = Protos.mcs_proto.HeartbeatAck;
    const errorMessage = HeartbeatAckRequestType.verify(heartbeatAckRequest);
    if (errorMessage) {
      throw new Error(errorMessage);
    }

    const buffer =
      HeartbeatAckRequestType.encodeDelimited(heartbeatAckRequest).finish();

    Logger.verbose("HEARTBEAT sending PONG", heartbeatAckRequest);

    this.socket.write(
      Buffer.concat([Buffer.from([MCSProtoTag.kHeartbeatAckTag]), buffer])
    );
  }

  private sendLogin() {
    const gcm = this.config.credentials.gcm;
    const LoginRequestType = Protos.mcs_proto.LoginRequest;
    const hexAndroidId = Long.fromString(gcm.androidId).toString(16);
    const loginRequest: mcs_proto.ILoginRequest = {
      adaptiveHeartbeat: false,
      authService: 2,
      authToken: gcm.securityToken,
      id: `chrome-${this.config.chromeVersion}`,
      domain: "mcs.android.com",
      deviceId: `android-${hexAndroidId}`,
      networkType: 1,
      resource: gcm.androidId,
      user: gcm.androidId,
      useRmq2: true,
      setting: [{ name: "new_vc", value: "1" }],
      clientEvent: [],
      // Id of the last notification received
      receivedPersistentId: this.config.persistentIds,
    };

    if (this.config.heartbeatIntervalMs) {
      loginRequest.heartbeatStat = {
        ip: "",
        timeout: true,
        intervalMs: this.config.heartbeatIntervalMs,
      };
    }

    const errorMessage = LoginRequestType.verify(loginRequest);
    if (errorMessage) {
      throw new Error(errorMessage);
    }

    const buffer = LoginRequestType.encodeDelimited(loginRequest).finish();

    this.socket.write(
      Buffer.concat([
        Buffer.from([Variables.kMCSVersion, MCSProtoTag.kLoginRequestTag]),
        buffer,
      ])
    );
  }

  private handleMessage = ({ tag, object }: Types.DataPacket): void => {
    // any message will reset the client side heartbeat timeout.
    this.startHeartbeat();

    switch (tag) {
      case MCSProtoTag.kLoginResponseTag:
        // clear persistent ids, as we just sent them to the server while logging in
        this.config.persistentIds = [];
        this.emit("ON_READY");
        this.startHeartbeat();
        break;

      case MCSProtoTag.kDataMessageStanzaTag:
        this.handleDataMessage(object);
        break;

      case MCSProtoTag.kHeartbeatPingTag:
        this.emit("ON_HEARTBEAT");
        Logger.verbose("HEARTBEAT PING", object);
        this.sendHeartbeatPong(object);
        break;

      case MCSProtoTag.kHeartbeatAckTag:
        this.emit("ON_HEARTBEAT");
        Logger.verbose("HEARTBEAT PONG", object);
        break;

      case MCSProtoTag.kCloseTag:
        Logger.verbose(
          "Close: Server requested close! message: ",
          JSON.stringify(object)
        );
        this.handleSocketClose();
        break;

      case MCSProtoTag.kLoginRequestTag:
        Logger.verbose("Login request: message: ", JSON.stringify(object));
        break;

      case MCSProtoTag.kIqStanzaTag:
        Logger.debug(
          "IqStanza: If anyone knows what is this and how to respond, please let me know! - message: ",
          JSON.stringify(object)
        );
        // FIXME: If anyone knows what is this and how to respond, please let me know
        break;

      default:
        Logger.error("Unknown message: ", JSON.stringify(object));
        return;

      // no default
    }

    this.streamId++;
  };

  private handleDataMessage = (object): void => {
    if (this.persistentIds.includes(object.persistentId)) {
      return;
    }

    let message;
    try {
      message = decrypt(object, this.config.credentials.keys);
    } catch (error) {
      console.error(error?.response);
      switch (true) {
        case error.message.includes(
          "Unsupported state or unable to authenticate data"
        ):
        case error.message.includes("crypto-key is missing"):
        case error.message.includes("salt is missing"):
          // NOTE(ibash) Periodically we're unable to decrypt notifications. In
          // all cases we've been able to receive future notifications using the
          // same keys. So, we silently drop this notification.
          Logger.warn(
            "Message dropped as it could not be decrypted: " + error.message
          );
          return;
        default:
          throw error;
      }
    }

    // Maintain persistentIds updated with the very last received value
    this.persistentIds.push(object.persistentId);
    // Send notification
    this.emit("ON_MESSAGE_RECEIVED", {
      message,
      // Needs to be saved by the client
      persistentId: object.persistentId,
    });
  };

  private handleParserError = (error) => {
    Logger.error(error);
    this.socketRetry();
  };

  public send(
    message: Types.MessageToSend,
    serverApiKey: string
  ): Promise<void> {
    Logger.verbose("testMessage");
    if (!serverApiKey)
      throw new Error("Can't test messages without serverApiKey");

    const data = {
      time_to_live: 3,
      data: message,
      registration_ids: [this.config.credentials.fcm.token],
    };

    return axios
      .post(FCM_SEND_API, data, {
        headers: {
          Authorization: `key=${serverApiKey}`,
        },
      })
      .then(({ data }) => {
        if (data.failure) {
          Logger.debug("Message test failed");
          throw new Error(data);
        }

        Logger.debug("Message test passed");
      });
  }

  public testMessage(serverApiKey: string): Promise<void> {
    Logger.verbose("testMessage");
    return this.send(
      {
        message: "PushReceiver test message",
        title: "testMessage",
        key: "",
        action: "",
      },
      serverApiKey
    );
  }
}

export { PushReceiver };
